cpp/src/arrow/dataset/scanner.h

```diff
@@ -138,47 +179,147 @@ ARROW_DS_EXPORT Result<ScanTaskIterator> ScanTaskIteratorFromRecordBatch(
     std::vector<std::shared_ptr<RecordBatch>> batches,
     std::shared_ptr<ScanOptions> options);
 
-/// \brief Scanner is a materialized scan operation with context and options
-/// bound. A scanner is the class that glues ScanTask, Fragment,
-/// and Dataset. In python pseudo code, it performs the following:
+template <typename T>
+struct Enumerated {
+  T value;
+  int index;
+  bool last;
+};
+
+/// \brief Combines a record batch with the fragment that the record batch originated
+/// from
+///
+/// Knowing the source fragment can be useful for debugging & understanding loaded data
+struct TaggedRecordBatch {
+  std::shared_ptr<RecordBatch> record_batch;
+  std::shared_ptr<Fragment> fragment;
+};
```

Comment:
Yes, that's why I ended up using visit in `FileSystemDataset::Write`.  However, the `Iterator` is exposed externally so it might be nice to add support for it.  Unfortunately, neither comparing record batches or comparing fragments is simple (consider in memory fragments) or cheap.

I could use reference equality again (I rely on fragments having stable references in `Scanner::AddPositioningToInOrderScan`) and that should work but I don't see it used much elsewhere so I doubt myself.

What do you think?

Reply:
Hmm, that's fair, it's probably best not to expose a performance trap to everyone. It's a little unfortunate that C++ iterators require full equality. Requiring Visit()/accepting that the interface will be a little hard to work with is OK by me since we can still expose good Python/R bindings regardless.